lib/searchkick/query.rb

```diff
@@ -805,70 +805,86 @@ def set_aggregations(payload, filters, post_filters)
+        filters.select! do |filter|
+          if agg_filters.include?(filter)
+            true
+          else
+            post_filters << filter
+            false
+          end
+        end
+
+        if agg_filters.any?
           payload[:aggs][field] = {
+            filter: {
+              bool: {
+                must: agg_filters
+              }
+            },
+            aggs: {
+              field => payload[:aggs][field]
+            }
+          }
+        end
+      end
+    end
+
+    def build_aggregation(field, agg_options)
+      size = agg_options[:limit] ? agg_options[:limit] : 1_000
+      shared_agg_options = agg_options.except(:aggs, :date_ranges, :field, :limit, :ranges, :where)
+
+      aggregation =
+        if agg_options[:ranges]
+          {
             range: {
               field: agg_options[:field] || field,
               ranges: agg_options[:ranges]
             }.merge(shared_agg_options)
           }
         elsif agg_options[:date_ranges]
-          payload[:aggs][field] = {
+          {
             date_range: {
               field: agg_options[:field] || field,
               ranges: agg_options[:date_ranges]
             }.merge(shared_agg_options)
           }
         elsif (histogram = agg_options[:date_histogram])
-          payload[:aggs][field] = {
+          {
             date_histogram: histogram
           }.merge(shared_agg_options)
         elsif (metric = @@metric_aggs.find { |k| agg_options.has_key?(k) })
-          payload[:aggs][field] = {
+          {
             metric => {
               field: agg_options[metric][:field] || field
             }
           }.merge(shared_agg_options)
         else
-          payload[:aggs][field] = {
+          {
             terms: {
               field: agg_options[:field] || field,
               size: size
             }.merge(shared_agg_options)
           }
         end
 
-        where = {}
-        where = ensure_permitted(options[:where] || {}).reject { |k| k == field } unless options[:smart_aggs] == false
-        agg_where = ensure_permitted(agg_options[:where] || {})
-        agg_filters = where_filters(where.merge(agg_where))
+      # nested aggregations
+      if agg_options[:aggs].present?
+        aggregation[:aggs] = {}
 
-        # only do one level comparison for simplicity
-        filters.select! do |filter|
-          if agg_filters.include?(filter)
-            true
-          else
-            post_filters << filter
-            false
-          end
-        end
-
-        if agg_filters.any?
-          payload[:aggs][field] = {
-            filter: {
-              bool: {
-                must: agg_filters
-              }
-            },
-            aggs: {
-              field => payload[:aggs][field]
-            }
-          }
+        agg_options[:aggs].each do |field, nested_agg_options|
+          aggregation[:aggs][field] = build_aggregation(field, nested_agg_options)
```

Comment:
In case of nested aggregations, we recursively call `build_aggregations` for each aggregation.
Lines 879 - 885 represent the main change in this PR.